consensus/safety-rules/src/safety_rules.rs

```diff
@@ -235,6 +235,42 @@ impl<T: Payload> TSafetyRules<T> for SafetyRules<T> {
     /// @TODO verify QC matches preferred round
     fn sign_proposal(&mut self, block_data: BlockData<T>) -> Result<Block<T>, Error> {
         debug!("Incoming proposal to sign.");
+        let proposal_epoch = block_data.epoch();
+        let proposal_round = block_data.round();
+        let last_voted_round = self.persistent_storage.last_voted_round()?;
+
+        // proposed epoch must match the current epoch
+        self.verify_epoch(proposal_epoch)?;
+
+        // proposed round must be increasing
+        if proposal_round <= last_voted_round {
+            debug!(
+                "Block round is older than last_voted_round ({} <= {})",
+                proposal_round, last_voted_round
+            );
+            return Err(Error::OldProposal {
```

Comment:
not related to this pr, but this pattern of println + return Err makes me sad, can we have some magic that both prints and return the error?

we probably want another error type for this  to differentiate self-propose old proposal vs receiving remote old proposal

Reply:
I think the error is okay, because the context is different... and the text for the error is pretty neutral: https://github.com/libra/libra/blob/master/consensus/safety-rules/src/error.rs#L44